### DISCUSSION TITLE
Resolved LES in Oceananigans

### DISCUSSION THREAD

**Discussion by mmr0** [CliMA/Oceananigans.jl#3129] [Computational science!]
I'm after some help setting up a resolved LES in Oceananigans. By resolved LES i'm specifically referring to resolving the viscous sublayer/wall stress for a wall-bounded flow, e.g. no wall model needed. Has anyone done this, and if so are you willing to share your experience?

Where I'm at: I'm currently trying to replicate a DIABLO simulation of an upside down bottom Ekman layer (i.e. the no-slip boundary is at the top of the domain) as a test but despite having the same resolution as I've used with DIABLO I don't have anything that looks like a viscous sublayer in my simulation, and the stress at the interface is also not what I expect. This was using `closure = AnisotropicMinimumDissipation()`. Code is below if interested. Note that the code runs and i get pretty sensible looking outer boundary layer and mixed layer, it's just the near-wall stuff that I'm a bit dubious of. 

A potentially related issue: I was looking into the Tuple closure method, which would suit me since I'd like to modify the molecular viscosity to something suitable for Polar oceans. I tried:
`amd_closure = (AnisotropicMinimumDissipation(), ScalarDiffusivity(ν=2.0e-6, κ=1.4e-7))`
but got the error `type Tuple has no field νₑ`. I saw that this was likely associated with a bug fixed a few versions ago by @tomcho, so I updated updated to the latest version, but I still get the error. 

A further potentially related issue: Also came across the issue #1879 regarding tests that the tuple closure method works, and wondered if these were ever carried out?

Sorry for the lengthy post! This is my first foray into Oceananigans, so I'm really not sure what's just user error at this point. Cheers!

```
using Random
using Printf
using CairoMakie
using Oceananigans
using Oceananigans.Units: minute, minutes, hour

Nz = 300          # number of points in the vertical direction
Lz = 6          # (m) domain depth

# Stretched grid using tanh
h(k) = (k - 1) / Nz
const CSY = 2.5
z_faces(k) = Lz * (tanh(CSY * h(k)) / tanh(CSY) - 1)

grid = RectilinearGrid(GPU();
                       size = (128, 128, Nz),
                          x = (0, 6),
                          y = (0, 6),
                          z = z_faces)

buoyancy = SeawaterBuoyancy(equation_of_state=LinearEquationOfState(thermal_expansion = 2e-4,
                                                                    haline_contraction = 8e-4))
dTdz = 0.0015 # K m⁻¹
T_bcs = FieldBoundaryConditions(top = GradientBoundaryCondition(0.0),
                                bottom = GradientBoundaryCondition(dTdz))
u_bcs = FieldBoundaryConditions(top = ValueBoundaryCondition(0.0))
v_bcs = FieldBoundaryConditions(top = ValueBoundaryCondition(0.0))
w_bcs = FieldBoundaryConditions(top = ValueBoundaryCondition(0.0))
S_bcs = FieldBoundaryConditions(top=GradientBoundaryCondition(0.0))

const u_m = 0.014
coriolis = FPlane(f=1.4e-4)

@inline geostrophy(x, y, z, t, p) = @inbounds -p.U0 * p.f
v_forcing = Forcing(geostrophy, parameters = (U0=-u_m, f=coriolis.f))

amd_closure = (AnisotropicMinimumDissipation(), ScalarDiffusivity(ν=2.0e-6, κ=1.4e-7))

model = NonhydrostaticModel(; grid, buoyancy,
                            advection = advection = UpwindBiasedFifthOrder(),
			    timestepper = :RungeKutta3,
                            tracers = (:T, :S),
                            coriolis = coriolis,
                            closure = amd_closure, #AnisotropicMinimumDissipation(),
                            boundary_conditions = (u=u_bcs,v=v_bcs, T=T_bcs, S=S_bcs),
			    forcing = (v=v_forcing,)
			    )

# Random noise damped at top and bottom
Ξ(z) = randn() * z / model.grid.Lz * (1 + z / model.grid.Lz) # noise

# Temperature initial condition: a stable density gradient with random noise superposed.
Tᵢ(x, y, z) = 20 + dTdz * z + dTdz * model.grid.Lz * 1e-6 * Ξ(z)

# Velocity initial condition - needs to satisfy u_bc and gesostrophy
u_init(z) = -u_m * tanh(z*3)

# Velocity initial condition: random noise scaled by the friction velocity.
uᵢ(x, y, z) = u_init(z) + u_m*1e-3 * Ξ(z)
wᵢ(x, y, z) = u_m * 1e-3 * Ξ(z)
# `set!` the `model` fields using functions or constants:
set!(model, u=uᵢ, w=wᵢ, T=Tᵢ, S=35)
```

**Comment by glwagner** [comment]
We believe this use case is important and so we explicitly test it:

https://github.com/CliMA/Oceananigans.jl/blob/2fb0f77f753d3d30710ac37a82e8cef70dfd7fcb/test/test_turbulence_closures.jl#L137-L145

in addition to some other tests for tuple closures. That said, there is probably room to expand the tests for tuple closures. But I'm not sure the tuple itself is your issue.

> I tried: amd_closure = (AnisotropicMinimumDissipation(), ScalarDiffusivity(ν=2.0e-6, κ=1.4e-7)) but got the error type Tuple has no field νₑ.

This error is probably associated with the code you are using for diagnostics. Can you share that?

**Reply by glwagner** [reply]
Secondly, to do wall-resolved LES you'll have to set `ValueBoundaryCondition(0)` for the eddy viscosity and eddy diffusivities. I'm not sure how many people attempt this, but for resolving the _bottom_ boundary layer (for example) I believe something like this will work:

```julia
# Eddy diffusivity boundary conditions
zero_bcs = ValueBoundaryCondition(0)
zero_bcs = FieldBoundaryConditions(bottom=ValueBoundaryCondition(0))
eddy_diffusivity_boundary_conditions = (νₑ = zero_bcs, κₑ = (T = zero_bcs, S = zero_bcs))

prognostic_boundary_conditions = (u=u_bcs, v=v_bcs, T=T_bcs, S=S_bcs)

boundary_conditions = merge(prognostic_boundary_conditions, eddy_diffusivity_boundary_conditions)
```

PS its ok to use `ValueBoundaryCondition(0)`. Using `0.0` doesn't change performance or results.

The eddy diffusivity boundary conditions make their way into the eddy diffusivity fields here:

https://github.com/CliMA/Oceananigans.jl/blob/2fb0f77f753d3d30710ac37a82e8cef70dfd7fcb/src/TurbulenceClosures/turbulence_closure_implementations/anisotropic_minimum_dissipation.jl#L341-L353

PPS Elsewhere we discussed making this a little more convenient, ie through an option specified when constructing `AnisotropicMinimumDissipation`. It's probably worth considering. @tomchor what do you think?

As a side note and for the future, I don't think we support wall-resolved LES with immersed boundaries. We'd have to make some modifications to how the immersed boundary condition is implemented to support that, I think. Food for thought.

**Reply by tomchor** [reply]
> This error is probably associated with the code you are using for diagnostics. Can you share that?

Agreed. It would be useful to see the whole code (and the animations/plots) to be able to see what's going on.

> PPS Elsewhere we discussed making this a little more convenient, ie through an option specified when constructing AnisotropicMinimumDissipation. It's probably worth considering. @tomchor what do you think?

Also agreed! With Oceananigans becoming ever more flexible the chances this option being useful to users increases, and I don't think it'd be too much extra code.

CC @zhazorken

**Reply by glwagner** [reply]
Maybe a boolean `zero_on_boundaries=true` or `wall_resolved=true`, which would then set `ValueBoundaryCondition(0)` on both sides of all `Bounded` directions, as well as on immersed boundaries.

**Reply by mmr0** [reply]
Thanks @glwagner and @tomchor for getting back to me. The rest of the code (that gives error `type Tuple has no field νₑ`) is below [Note that I don't get that error with `closure = AnisotropicMinimumDissipation()`, just when I try the tuple closure]. 

If I comment out the eddy viscosity output bits the code runs but exits straight away due to Nans in all fields. 

```
simulation = Simulation(model, Δt=2.0, stop_time=360minutes)

wizard = TimeStepWizard(cfl=1.0, max_change=1.1, max_Δt=1minute)
simulation.callbacks[:wizard] = Callback(wizard, IterationInterval(10))

# Print a progress message
progress_message(sim) = @printf("Iteration: %04d, time: %s, Δt: %s, max(|w|) = %.1e ms⁻¹, wall time: %s\n",
                                iteration(sim), prettytime(sim), prettytime(sim.Δt),
                                maximum(abs, sim.model.velocities.w), prettytime(sim.run_wall_time))

simulation.callbacks[:progress] = Callback(progress_message, IterationInterval(20))

# Create a NamedTuple with eddy viscosity
eddy_viscosity = (; νₑ = model.diffusivity_fields.νₑ)

filename = "test_2closure"

simulation.output_writers[:slices] =
    NetCDFOutputWriter(model, merge(model.velocities, model.tracers, eddy_viscosity),
                     filename = filename * "_planes.nc",
                     indices = (:, grid.Ny/2, :),
                     schedule = TimeInterval(10minute),
                     overwrite_existing = true)

# Write out plane-averaged profiles
T_avg = Field(Average(model.tracers.T, dims=(1, 2)))
S_avg = Field(Average(model.tracers.S, dims=(1, 2)))
u_avg = Field(Average(model.velocities.u, dims=(1,2)))
v_avg = Field(Average(model.velocities.v, dims=(1,2)))

outputs = Dict("T" => T_avg, "S" => S_avg,
               "u" => u_avg, "v" => v_avg);
dims = Dict("T" => ("zC",), "S" => ("zC",),
            "u" => ("zC",), "v" => ("zC",));

simulation.output_writers[:xy_averaged_profile_writer] =
        NetCDFOutputWriter(model, outputs,
        filename = filename * "_profiles.nc",
        dimensions = dims,
        schedule = TimeInterval(1minute),
        overwrite_existing = true)

run!(simulation)
```

**Reply by mmr0** [reply]
> Agreed. It would be useful to see the whole code (and the animations/plots) to be able to see what's going on.

Copy! Here are some plots of the comparison with DIABLO for the the case with `closure = AnisotropicMinimumDissipation()` which runs fine but looks a bit odd: 
Grid res and velocity profiles (horizontally averaged):
![vel_prof_comparison_001](https://github.com/CliMA/Oceananigans.jl/assets/17136023/7fe52375-a878-44cf-8bc3-6c5e5ca40797)
the semilog plot best shows the viscous sublayer and log layer in DIABLO, but not Oceananigans. 
Drag coefficient $C_d = (u^*/U)^2$ and stratification:
![drag_timeseries_001](https://github.com/CliMA/Oceananigans.jl/assets/17136023/4377efef-d0f4-451b-8851-e02570d46480)
Planes [Oceananigans only]:
![planes_Oc_001](https://github.com/CliMA/Oceananigans.jl/assets/17136023/83288ad4-578c-4a0e-bc74-01a190d76fbe)

**Reply by mmr0** [reply]
> Secondly, to do wall-resolved LES you'll have to set ValueBoundaryCondition(0) for the eddy viscosity and eddy diffusivities. I'm not sure how many people attempt this, but for resolving the bottom boundary layer (for example) I believe something like this will work:

```
# Eddy diffusivity boundary conditions
zero_bcs = ValueBoundaryCondition(0)
zero_bcs = FieldBoundaryConditions(bottom=ValueBoundaryCondition(0))
eddy_diffusivity_boundary_conditions = (νₑ = zero_bcs, κₑ = (T = zero_bcs, S = zero_bcs))

prognostic_boundary_conditions = (u=u_bcs, v=v_bcs, T=T_bcs, S=S_bcs)

boundary_conditions = merge(prognostic_boundary_conditions, eddy_diffusivity_boundary_conditions)
```

Ok thanks @glwagner I will try this. Can you comment on what the first two lines are doing (zero_bcs twice?)

Cheers

**Reply by tomchor** [reply]
In this line

```julia
eddy_viscosity = (; νₑ = model.diffusivity_fields.νₑ)
```

you're trying to access `model.diffusivity_fields.νₑ` manually, which indeed doesn't exist when you're using a Tuple closure. To get the eddy viscosity you should use:

```julia
eddy_viscosity = (; νₑ = viscosity(model.closure, model.diffusivity_fields))
```

PS: I haven't tested the line above so it's possible I inverted the order of the arguments :grimacing: , but that should be it!

**Reply by glwagner** [reply]
> Can you comment on what the first two lines are doing (zero_bcs twice?)

Sorry... that was a typo. You only need 

```julia
zero_bcs = FieldBoundaryConditions(bottom=ValueBoundaryCondition(0))
```

**Reply by glwagner** [reply]
Also, you need to set this at the boundary you're trying to resolve. To be safe you can do both the top and the bottom (from the plots above it looks like you want to resolved the top boundary layer):

```julia
zero_bcs = FieldBoundaryConditions(top=ValueBoundaryCondition(0), bottom=ValueBoundaryCondition(0))
```

**Reply by mmr0** [reply]
I'm still working through a couple of things (e.g. a good initial condition to kick-off the turbulence from the get-go), but looking pretty good! Thanks for your help so far @tomchor @glwagner 

![vel_prof_comparison_001 (1)](https://github.com/CliMA/Oceananigans.jl/assets/17136023/6af92e05-7f66-4544-abc2-aad8b7959904)
![drag_timeseries_001](https://github.com/CliMA/Oceananigans.jl/assets/17136023/d315149f-a317-4b6e-b9be-ab28e9d60491)

**Reply by glwagner** [reply]
Nice. Since the numerics are different, maybe we also need to test multiple resolutions?

I'd be curious about the effect of different combinations of LES closures and advection schemes. For example, we usually run LES with `advection = WENO()` and the `AnisotropicMinimumDissipation()` closure. However, it would be also interesting to test `advection = CenteredSecondOrder()` (arguably closer to what DIABLO uses), and also `advection = WENO()` with no LES closure (ie just `closure = ScalarDiffusivity()`).

**Reply by glwagner** [reply]
Are you running with `UpwindBiasedFifthOrder()`? If so I recommend switching to WENO as a first step.

**Reply by mmr0** [reply]
Yes I recently switched to WENO on @zhazorken recommendation! Looks similar at first glance. I will investigate some different combinations and report back!

**Reply by glwagner** [reply]
Another thing that might be important if you have very fine grid spacing near the wall is `VerticallyImplicitTimeDiscretization`:

```julia
vitd = VerticallyImplicitTimeDiscretization()
molecular_diffusivity = ScalarDiffusivity(vitd, ...)
```

you may also want to use for AMD.

**Reply by mmr0** [reply]
Thanks @glwagner - I am just coming back to this now so great timing!

**Comment by navidcy** [comment]
Oh I just saw this! Nice to see you started running Oceananigans @mmr0 !
